README.md
Scrypt.d
========

A library for using Scrypt encryption in the D programming language

Implemented as a DUB package (optionally buildable as a regular D library)

Features
--------

Scrypt.d comes with bindings for crypto_scrypt.h and scryptenc.h which means you can use this to either encode or decode
data/files or use it to hash data (one way).

It also comes with it's own wrapper around crypto_scrypt to make it easier to hash passwords in D, there are instructions on how to do that right below!

### Using the scrypt.password module

Because you don't wanna deal with typecasting and other bullshit, I decided to make it easy to hash your goddamned passwords.

password.d comes with 3 functions:

``` d
string genRandomSalt();
string genScryptPasswordHash(string password, string salt, size_t scrypt_outputlen, ulong N, uint r, uint p);
bool checkScryptPasswordHash(string hash, string password);
```

Okay so password is obviously the parameter for the password you either wanna check or hash, salt is the random salt
that you want to use in the hashing process (passing genRandomSalt() should be fine), and hash is the already hashed
version of your password that you want to check against in checkScryptPasswordHash.

Now where it gets tricky is when you see the scrypt_outputlen, N, r and p parameters. What the fuck are those things?
Let me explain:

scrypt_outputlen is basically the size of the buffer that scrypt outputs to, it isn't the length of the final hash or
anything, since that's a sha1 digest of the hash prepended by the salt.

N is the General work factor. This should be a power of two. The suggested one for passwords is currently 2^14 (but if you want to use SCRYPT_N_DEFAULT, although I wouldn't trust an external library not to change that value so I'd just put my own constant in the code, do as you like, I will probably not change it ever though)
2^20 is what you should use for sensitive stuff atm e.g credit card numbers.

r is the blocksize used for the hash. Currently 8 seems to be fine (or SCRYPT_R_DEFAULT)

and p is the parallelization factor, 1 is currently a nice default. (or SCRYPT_P_DEFAULT)

All of the confusing params (scrypt_outputlen, N, r and p) *must* and I repeat __MUST__ be the same when you are calling
checkScryptPasswordHash as they were when you where calling genScryptPasswordHash, othewise they simply won't be tha same hashes.

Now here is an example of using the password module for great win:

``` d
import scrypt.password;

// Some function for saving a password in to your database for example
string password = genScryptPasswordHash(input_password_string, genRandomSalt(), SCRYPT_OUTPUTLEN_DEFAULT, SCRYPT_N_DEFAULT, SCRYPT_R_DEFAULT, SCRYPT_P_DEFAULT);
db.save("foo", password);
// end

// Some function pulling out the data from the db to check if the password matches the one you saved earlier
string password_hash = db.get("foo");
bool authenticate = checkScryptPasswordHash(password_hash, input_password);
// end
```

Simple right? Yes it's a bit wordy, but I'm giving YOU full control here remember? Now go store passwords securely! No insecure md5 or sha1 hashes, I'll smack you down!

Dependencies
------------

+ Tarsnap (you'll need to build scrypt as a library, see "Build Instructions" below)

Build Instructions
------------------

#### POSIX instructions (this means linux, freebsd etc)

> The way I got this to run properly on my machine was to download and compile tarsnap (found at https://www.tarsnap.com/download.html)
> I then grabbed the libtarsnap.a and put it where my compiler would find it e.g (/usr/local/lib) and after that it's done as tarsnap is listed in
> the package.json file as a dependency!

#### Windows instructions

> Someone else will have to write this as windows ain't worth my time. Should be similar to POSIX


CHANGELOG
0.1.0

+ Remove the sha1 digest of the password hash.
+ Store work factors in the hash string.

0.0.9

Works! But don't use this version
It makes a sha1 digest which opens up the possibility
that there will be hash collisions. Also it doesn't
return the work factor which you should store.


source/app.d
import std.stdio;
import scrypt.password;
import std.uuid : randomUUID;

/// Just try stuff out
void main() {
    writeln("example: ", genScryptPasswordHash("password", genRandomSalt(), SCRYPT_OUTPUTLEN_DEFAULT, SCRYPT_N_DEFAULT, SCRYPT_R_DEFAULT, SCRYPT_P_DEFAULT)); 
    writeln(genScryptPasswordHash("pw", genRandomSalt(), 90, SCRYPT_N_DEFAULT, 8, 1) != genScryptPasswordHash("pw", genRandomSalt(), 90, SCRYPT_N_DEFAULT, 8, 1));

    foreach(i; 1 .. 10)
        writeln(genScryptPasswordHash("pw", genRandomSalt(), 90, SCRYPT_N_DEFAULT, 8, 1));

    foreach(i; 1 .. 5) {
        string pw = genScryptPasswordHash("very password nice really, random kind the you use should really", genRandomSalt(), 90, SCRYPT_N_DEFAULT, 8, 1);
        writeln(checkScryptPasswordHash(pw, "very password nice really, random kind the you use should really"));
    }

    // writeln("\nRunning huuuge test, stay put!\n");
    // runTest(); // Weird test
}

/// Testing stuff

/// Epic test on storing a bunch of users, not sure why I did this.
void runTest() {
    int usercount = 1000;

    writeln("Creating usernames...");

    string[] userNames;
    foreach(i; 0 .. usercount) {
        userNames ~= randomUUID().toString();
    }

    writeln("Creating passwords...");
    string[] passwords;
    foreach(i; 0 .. usercount) {
        passwords ~= randomUUID().toString();
    }

    writeln("Creating users...");
    string[string] users;
    foreach(i; 0 .. usercount) {
        users[userNames[i]] = genScryptPasswordHash(passwords[i], genRandomSalt(), SCRYPT_OUTPUTLEN_DEFAULT, SCRYPT_N_DEFAULT, SCRYPT_R_DEFAULT, SCRYPT_P_DEFAULT);
    }
    
    writeln("Verifying every damn user...");
    bool allworked = true;
    foreach(i; 0 .. usercount) {
        writeln("Testing: ", i+1);
        if (!checkScryptPasswordHash(users[userNames[i]], passwords[i])) {
            allworked = false;
            break;
        }
    }

    if (allworked) {
        writeln("Huzzaaaaahhhh!");
    } else {
        writeln("WTF??!?!?! ;______;");
    }
}


source/scrypt/scryptenc.d
module scrypt.scryptenc;
/*-
 * Copyright 2009 Colin Percival
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * This file was originally written by Colin Percival as part of the Tarsnap
 * online backup system.
 * 
 * D binding created by Isak Andersson 2013 (BitPuffin@lavabit.com)
 */

import std.c.stdio;

alias ubyte uint8_t;

extern (C):

/**
 * The parameters maxmem, maxmemfrac, and maxtime used by all of these
 * functions are defined as follows:
 * maxmem - maximum number of bytes of storage to use for V array (which is
 *     by far the largest consumer of memory).  If this value is set to 0, no
 *     maximum will be enforced; any other value less than 1 MiB will be
 *     treated as 1 MiB.
 * maxmemfrac - maximum fraction of available storage to use for the V array,
 *     where "available storage" is defined as the minimum out of the
 *     RLIMIT_AS, RLIMIT_DATA. and RLIMIT_RSS resource limits (if any are
 *     set).  If this value is set to 0 or more than 0.5 it will be treated
 *     as 0.5; and this value will never cause a limit of less than 1 MiB to
 *     be enforced.
 * maxtime - maximum amount of CPU time to spend computing the derived keys,
 *     in seconds.  This limit is only approximately enforced; the CPU
 *     performance is estimated and parameter limits are chosen accordingly.
 * For the encryption functions, the parameters to the scrypt key derivation
 * function are chosen to make the key as strong as possible subject to the
 * specified limits; for the decryption functions, the parameters used are
 * compared to the computed limits and an error is returned if decrypting
 * the data would take too much memory or CPU time.
 */
/**
 * Return codes from scrypt(enc|dec)_(buf|file):
 * 0	success
 * 1	getrlimit or sysctl(hw.usermem) failed
 * 2	clock_getres or clock_gettime failed
 * 3	error computing derived key
 * 4	could not read salt from /dev/urandom
 * 5	error in OpenSSL
 * 6	malloc failed
 * 7	data is not a valid scrypt-encrypted block
 * 8	unrecognized scrypt format
 * 9	decrypting file would take too much memory
 * 10	decrypting file would take too long
 * 11	password is incorrect
 * 12	error writing output file
 * 13	error reading input file
 */

/**
 * scryptenc_buf(inbuf, inbuflen, outbuf, passwd, passwdlen,
 *     maxmem, maxmemfrac, maxtime):
 * Encrypt inbuflen bytes from inbuf, writing the resulting inbuflen + 128
 * bytes to outbuf.
 */
int scryptenc_buf(const uint8_t *, size_t, uint8_t *,
    const uint8_t *, size_t, size_t, double, double);

/**
 * scryptdec_buf(inbuf, inbuflen, outbuf, outlen, passwd, passwdlen,
 *     maxmem, maxmemfrac, maxtime):
 * Decrypt inbuflen bytes from inbuf, writing the result into outbuf and the
 * decrypted data length to outlen.  The allocated length of outbuf must
 * be at least inbuflen.
 */
int scryptdec_buf(const uint8_t *, size_t, uint8_t *, size_t *,
    const uint8_t *, size_t, size_t, double, double);

/**
 * scryptenc_file(infile, outfile, passwd, passwdlen,
 *     maxmem, maxmemfrac, maxtime):
 * Read a stream from infile and encrypt it, writing the resulting stream to
 * outfile.
 */
int scryptenc_file(FILE *, FILE *, const uint8_t *, size_t,
    size_t, double, double);

/**
 * scryptdec_file(infile, outfile, passwd, passwdlen,
 *     maxmem, maxmemfrac, maxtime):
 * Read a stream from infile and decrypt it, writing the resulting stream to
 * outfile.
 */
int scryptdec_file(FILE *, FILE *, const uint8_t *, size_t,
    size_t, double, double);



source/scrypt/crypto_scrypt.d
module scrypt.crypto_scrypt;

/*-
 * Copyright 2009 Colin Percival
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * This file was originally written by Colin Percival as part of the Tarsnap
 * online backup system.
 * 
 * D binding created by Isak Andersson 2013 (BitPuffin@lavabit.com)
 */

alias ubyte uint8_t;
alias ulong uint64_t;
alias uint uint32_t;

extern (C):

/**
 * crypto_scrypt(passwd, passwdlen, salt, saltlen, N, r, p, buf, buflen):
 * Compute scrypt(passwd[0 .. passwdlen - 1], salt[0 .. saltlen - 1], N, r,
 * p, buflen) and write the result into buf.  The parameters r, p, and buflen
 * must satisfy r * p < 2^30 and buflen <= (2^32 - 1) * 32.  The parameter N
 * must be a power of 2 greater than 1.
 *
 * Return 0 on success; or -1 on error.
 */
int crypto_scrypt(const uint8_t *, size_t, const uint8_t *, size_t, uint64_t,
    uint32_t, uint32_t, uint8_t *, size_t);




source/scrypt/password.d
module scrypt.password;

/*
 * Copyright (C) 2013 Isak Andersson (BitPuffin@lavabit.com)
 * 
 * Distributed under the terms of the zlib/libpng license
 * See LICENSE.txt in project root for more info
 */

import scrypt.crypto_scrypt;
import std.string : indexOf;
import std.exception : enforce;
import std.digest.digest : toHexString;
import std.uuid : randomUUID;
import std.algorithm : splitter;
import std.array: array;
import std.conv: to;

enum SCRYPT_N_DEFAULT = 16384;
enum SCRYPT_R_DEFAULT = 8;
enum SCRYPT_P_DEFAULT = 1;
enum SCRYPT_OUTPUTLEN_DEFAULT = 90;
private enum TERMINATOR = '$'; 

/// Takes no parameters, returns a random UUID in string form
string genRandomSalt() {
    return randomUUID().toString();
}

/**
  * Some info regarding the params
  * password: The password you want to hash, for example "my password"
  * salt: The salt you want to use when hashing your password, for example generateRandomSalt();
  * scrypt_outputlen: the length of the output string containing your hashed password from scrypt. Reccomended value is 90. Note, the actual output won't be 90 since it's a sha1 digest
  * N: General work factor, iteration count. Must be power of two. Recommended value for passwords: 2^14 and 2^20 for sensitive stuff
  * r: Blocksize for underlying hash. Reccommended value is 8
  * p: parallelization factor. Reccomended value is 1
  * If you want to you can use SCRYPT_N_DEFAULT, SCRYPT_R_DEFAULT, SCRYPT_P_DEFAULT, SCRYPT_OUTPUTLEN_DEFAULT as default params
  */
string genScryptPasswordHash(string password, string salt, size_t scrypt_outputlen, ulong N, uint r, uint p) {
    ubyte[] outpw = new ubyte[scrypt_outputlen];
    crypto_scrypt(cast(ubyte*)password.ptr, password.length, cast(ubyte*)salt.ptr, salt.length, N, r, p, outpw.ptr, outpw.length);
    
    return toHexString(outpw).idup ~ TERMINATOR ~ salt ~ TERMINATOR ~ to!string(scrypt_outputlen) ~ TERMINATOR ~ to!string(N) ~ TERMINATOR ~ to!string(r) ~ TERMINATOR ~ to!string(p);
}

/**
  * Some info regarding the params
  * hash: An already hashed version of your password, for example fetched from a database
  * password: The password you wish to see if it matches
  */
bool checkScryptPasswordHash(string hash, string password) {
    auto params = hash.splitter(TERMINATOR).array[1 .. $];
    enforce(params.length == 5, "invalid hash string, does not meet requirements");
    return genScryptPasswordHash(password, params[0], to!size_t(params[1]), to!ulong(params[2]), to!uint(params[3]), to!uint(params[4])) == hash;
}

